### PR TITLE
UserValueResolver and SecurityUserValueResolver improvement

### PR DESCRIPTION
**Pull request by davidkmenta**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | -
| License       | MIT
| Doc PR        | -

UserValueResolver and SecurityUserValueResolver should also resolve class implementing UserInterface.

### CODE REVIEW

File: src/Symfony/Bundle/SecurityBundle/SecurityUserValueResolver.php [hunk 1]
```diff
@@ -40,7 +40,7 @@ public function __construct(TokenStorageInterface $tokenStorage)
     public function supports(Request $request, ArgumentMetadata $argument)
     {
         // only security user implementations are supported
-        if (UserInterface::class !== $argument->getType()) {
+        if (!$argument->getType() || !$this->implementsCorrectInterface($argument->getType())) {
```
Comment (chalasr) [inline src/Symfony/Bundle/SecurityBundle/SecurityUserValueResolver.php]: Should be in `Symfony\Component\Security\Http\Controller\UserValueResolver` instead (see deprecation notice above)

File: src/Symfony/Bundle/SecurityBundle/SecurityUserValueResolver.php [hunk 2]
```diff
@@ -59,4 +59,14 @@ public function resolve(Request $request, ArgumentMetadata $argument)
...
     }
+
+    /**
+     * @param string $type
+     *
+     * @return bool
+     */
+    private function implementsCorrectInterface($type)
+    {
+        return UserInterface::class === $type || array_key_exists(UserInterface::class, class_implements($type));
```
Comment (chalasr) [inline src/Symfony/Bundle/SecurityBundle/SecurityUserValueResolver.php]: Let's inline this logic in the calling condition

**Review by chalasr — COMMENTED** [review]
this was on my list :)

File: src/Symfony/Bundle/SecurityBundle/SecurityUserValueResolver.php (hunk 1 above)
Comment (davidkmenta) [reply src/Symfony/Bundle/SecurityBundle/SecurityUserValueResolver.php]: It's in both classes

Comment (davidkmenta) [reply src/Symfony/Bundle/SecurityBundle/SecurityUserValueResolver.php]: It's in both classes :)

Comment (chalasr) [reply src/Symfony/Bundle/SecurityBundle/SecurityUserValueResolver.php]: Deprecated classes are frozen until they are suppressed, it should not get new features.

File: src/Symfony/Bundle/SecurityBundle/SecurityUserValueResolver.php (hunk 2 above)
Comment (nicolas-grekas) [reply src/Symfony/Bundle/SecurityBundle/SecurityUserValueResolver.php]: isn't this what `is_a()` or `is_subclass_of()` does?

**Comment by davidkmenta** [comment]
@iltar thanks for the link. I agree that it may collide with the ParamConverter. On the other hand I still think the UserValueResolver should support classes implementing the UserInterface.

Few quick ideas:
- What about making it configurable or togglable?
- Is a ParamConverter called before or after a ValueResolver? Or not at all?

File: src/Symfony/Bundle/SecurityBundle/SecurityUserValueResolver.php (hunk 2 above)
Comment (davidkmenta) [reply src/Symfony/Bundle/SecurityBundle/SecurityUserValueResolver.php]: Didn't know 😳thanks!

**Comment by jvasseur** [comment]
I agree with @davidkmenta here the param converters are executed before the value resolver so the concern highlighted in https://github.com/symfony/symfony/pull/18510#issuecomment-208946979 shouldn't be a problem.

Forcing users to type they controller arguments with `UserInterface` gives bad DX for IDE auto-completion/static analysis.

**Comment by iltar** [comment]
@davidkmenta @jvasseur If you type-hint an entity, it will actually try to inject that entity based on the request, which will fail, because there's no way of fetching it. This is _because_ the parameter converters are triggered first.

> Forcing users to type they controller arguments with `UserInterface` gives bad DX for IDE auto-completion/static analysis.

If people use it as quick way to get their current user object, why not provide a (by the SecurityBundle) configurable AVR that can return a given object based on the current `UserInterface`? That would be a generic solution to a common problem.

**Comment by jvasseur** [comment]
@iltar didn't think of that since I tend do disable the automatic conversion.

We need to think of the future of the future of param converters (https://github.com/sensiolabs/SensioFrameworkExtraBundle/pull/436 ?) maybe it's time to deprecate them and introduce configurable argument resolvers, this would allow us to have a configurable security user resolver.

**Comment by iltar** [comment]
Yes, _something_ has to be done about the parameter converters, but I'm not sure what and I don't have much time myself :cry: 

Regarding the current AVR, what if we add an additional interface to the `UserProviderInterface`?

```php
interface DomainUserProviderInterface
{
    // would return your entity for example, names to be discussed
    public function convertToDomainUser(UserInterface $user);
}
```
```yml
# follows the basic configuration
security
    providers:
        hostnet:
            id: App\Security\Authentication\CustomerPortal\UserProvider
```
This would make it possible for the AVR to resolve the security user to a domain user. If you implement the interface on your user provider, it will simply be "enabled" for whatever has that user. We have the firewall information available, so perhaps it's not even hard to find the user provider.

**Comment by davidkmenta** [comment]
So I guess this PR can be closed for now :)

**Comment by gmponos** [comment]
TBH I do not follow why this was closed because of bundle that it's not inside symfony.. Personally I do not use Sensio bundles... :(

I believe that this is a problem of sensio bundle that should either decorate the user AVR or overwrite it.

**Comment by linaori** [comment]
Symfony needs a generic controller configuration system that works similar to the idea behind the annotations from that bundle (but configurable without annotations as well). When this is done, AVR can fully replace the parameter converters and it won't be an issue anymore. 

The issue arises because *a lot* of developers *do* use the `SensioFrameworkExtraBundle`, because it also adds features such as `@Security`. There's also effort being made to deprecate the user object from the Token, so the idea is to be able to inject your Domain user into an action based on this token, which means this AVR wouldn't even be necessary anymore. This won't solve the aforementioned issue though.

**Pull request closed without merging**